File: frontend/src/services/api/space-cleanup.ts
/**
 * 空间清理 API 服务
 *
 * 改造说明:
 * 1. 移除 .data.data 模式，后端现在直接返回数据
 * 2. 添加 i18n 字段支持
 */

import axios from './axios'
import { I18nDict } from './types'

// 资源类型枚举
export enum ResourceType {
  USER_UPLOADS = 'user_uploads',
  SANDBOX_SHARED = 'sandbox_shared',
  SANDBOX_PIP_CACHE = 'sandbox_pip_cache',
  SANDBOX_PACKAGES = 'sandbox_packages',
  PLUGIN_DYNAMIC_PACKAGES = 'plugin_dynamic_packages',
  PROMPT_LOGS = 'prompt_logs',
  PROMPT_ERROR_LOGS = 'prompt_error_logs',
  NAPCAT_TEMP = 'napcat_temp',
  PLUGIN_DATA = 'plugin_data',
  APP_LOGS = 'app_logs',
  OTHER_DATA = 'other_data',
}

// 扫描状态枚举
export enum ScanStatus {
  IDLE = 'idle',
  SCANNING = 'scanning',
  COMPLETED = 'completed',
  FAILED = 'failed',
}

// 清理状态枚举
export enum CleanupStatus {
  PENDING = 'pending',
  RUNNING = 'running',
  COMPLETED = 'completed',
  FAILED = 'failed',
}

// 文件信息
export interface FileInfo {
  relative_path: string
  name: string
  size: number
  created_time: number
  modified_time: number
  chat_key?: string
  plugin_key?: string
}

// 聊天资源信息
export interface ChatResourceInfo {
  chat_key: string
  chat_name?: string
  total_size: number
  file_count: number
  files: FileInfo[]
}

// 插件资源信息
export interface PluginResourceInfo {
  plugin_key: string
  plugin_name?: string
  total_size: number
  file_count: number
}

// 资源分类
export interface ResourceCategory {
  resource_type: ResourceType
  display_name: string
  description: string
  total_size: number
  file_count: number
  can_cleanup: boolean
  risk_level: string
  risk_message?: string
  supports_time_filter: boolean
  chat_resources: ChatResourceInfo[]
  plugin_resources: PluginResourceInfo[]

  // i18n 扩展字段
  i18n_description?: I18nDict
  i18n_display_name?: I18nDict
  i18n_risk_message?: I18nDict
}

// 磁盘信息
export interface DiskInfo {
  total_space: number
  used_space: number
  free_space: number
  data_dir_size: number
  data_dir_path: string
}

// 扫描摘要
export interface ScanSummary {
  total_size: number
  total_files: number
  start_time?: string
  end_time?: string
  duration_seconds?: number
  scanned_categories: number
}

// 扫描结果
export interface ScanResult {
  scan_id: string
  status: ScanStatus
  disk_info?: DiskInfo
  categories: ResourceCategory[]
  summary: ScanSummary
  error_message?: string
}

// 扫描状态响应
export interface ScanStatusResponse {
  status: ScanStatus
  progress: number
  current_category?: string
  message?: string
}

// 清理请求
export interface CleanupRequest {
  resource_types: ResourceType[]
  chat_keys?: string[]
  before_date?: string
  dry_run?: boolean
}

// 清理进度
export interface CleanupProgress {
  task_id: string
  status: CleanupStatus
  progress: number
  processed_files: number
  total_files: number
  freed_space: number
  current_file?: string
  message?: string
}

// 清理结果
export interface CleanupResult {
  task_id: string
  status: CleanupStatus
  total_files: number
  deleted_files: number
  failed_files: number
  freed_space: number
  start_time?: string
  end_time?: string
  duration_seconds?: number
  error_message?: string
  failed_file_list: string[]
}

// 简单响应类型（与后端 Pydantic 模型对应）
export interface ScanStartResponse {
  scan_id: string
}

export interface CleanupStartResponse {
  task_id: string
}

export interface ScanProgressResponse {
  status: string
  progress: number
  message?: string
}

export const spaceCleanupApi = {
  // 启动扫描
  startScan: async (): Promise<ScanStartResponse> => {
    const response = await axios.post<ScanStartResponse>('/space-cleanup/scan/start')
    return response.data
  },

  // 获取扫描状态
  getScanStatus: async (): Promise<ScanStatusResponse> => {
    const response = await axios.get<ScanStatusResponse>('/space-cleanup/scan/status')
    return response.data
  },

  // 获取扫描结果
  getScanResult: async (): Promise<ScanResult> => {
    const response = await axios.get<ScanResult>('/space-cleanup/scan/result')
    return response.data
  },

  // 启动清理
  startCleanup: async (request: CleanupRequest): Promise<CleanupStartResponse> => {
    const response = await axios.post<CleanupStartResponse>(
      '/space-cleanup/cleanup/start',
      request
    )
    return response.data
  },

  // 获取清理进度
  getCleanupProgress: async (taskId: string): Promise<CleanupProgress> => {
    const response = await axios.get<CleanupProgress>(
      `/space-cleanup/cleanup/progress/${taskId}`
    )
    return response.data
  },

  // 获取清理结果
  getCleanupResult: async (taskId: string): Promise<CleanupResult> => {
    const response = await axios.get<CleanupResult>(`/space-cleanup/cleanup/result/${taskId}`)
    return response.data
  },

  // 获取磁盘信息
  getDiskInfo: async (): Promise<DiskInfo> => {
    const response = await axios.get<DiskInfo>('/space-cleanup/disk-info')
    return response.data
  },

  // 从缓存加载扫描结果
  loadScanResultFromCache: async (): Promise<ScanResult> => {
    const response = await axios.get<ScanResult>('/space-cleanup/scan/load-cache')
    return response.data
  },

  // 获取扫描进度
  getScanProgress: async (): Promise<ScanProgressResponse> => {
    const response = await axios.get<ScanProgressResponse>('/space-cleanup/scan/progress')
    return response.data
  },
}

// 格式化字节大小
export function formatBytes(bytes: number, decimals: number = 2): string {
  if (bytes === 0) return '0 B'

  const k = 1024
  const dm = decimals < 0 ? 0 : decimals
  const sizes = ['B', 'KB', 'MB', 'GB', 'TB', 'PB']

  const i = Math.floor(Math.log(bytes) / Math.log(k))

  return parseFloat((bytes / Math.pow(k, i)).toFixed(dm)) + ' ' + sizes[i]
}

// 格式化时长（支持 i18n）
export function formatDuration(
  seconds: number,
  t?: (key: string, options?: Record<string, unknown>) => string
): string {
  // 如果没有传入 t 函数，使用默认中文
  if (!t) {
    if (seconds < 60) {
      return `${seconds.toFixed(1)} 秒`
    } else if (seconds < 3600) {
      const minutes = Math.floor(seconds / 60)
      const secs = Math.floor(seconds % 60)
      return `${minutes} 分 ${secs} 秒`
    } else {
      const hours = Math.floor(seconds / 3600)
      const minutes = Math.floor((seconds % 3600) / 60)
      return `${hours} 小时 ${minutes} 分`
    }
  }

  // 使用 i18n
  if (seconds < 60) {
    return t('spaceCleanup.duration.seconds', { seconds: seconds.toFixed(1) })
  } else if (seconds < 3600) {
    const minutes = Math.floor(seconds / 60)
    const secs = Math.floor(seconds % 60)
    return t('spaceCleanup.duration.minutesSeconds', { minutes, seconds: secs })
  } else {
    const hours = Math.floor(seconds / 3600)
    const minutes = Math.floor((seconds % 3600) / 60)
    return t('spaceCleanup.duration.hoursMinutes', { hours, minutes })
  }
}

File: nekro_agent/schemas/space_cleanup.py
"""空间清理相关数据模型"""

from datetime import datetime
from enum import Enum
from typing import Dict, List, Optional

from pydantic import BaseModel, Field

from nekro_agent.schemas.i18n import I18nDict


class ResourceType(str, Enum):
    """资源类型枚举"""

    USER_UPLOADS = "user_uploads"  # 用户上传资源
    SANDBOX_SHARED = "sandbox_shared"  # 沙盒共享目录
    SANDBOX_PIP_CACHE = "sandbox_pip_cache"  # pip缓存
    SANDBOX_PACKAGES = "sandbox_packages"  # 沙盒包
    PLUGIN_DYNAMIC_PACKAGES = "plugin_dynamic_packages"  # 插件动态包
    PROMPT_LOGS = "prompt_logs"  # 提示词日志
    PROMPT_ERROR_LOGS = "prompt_error_logs"  # 错误日志
    NAPCAT_TEMP = "napcat_temp"  # NapCat临时文件
    PLUGIN_DATA = "plugin_data"  # 插件数据（按插件分组）
    APP_LOGS = "app_logs"  # 应用日志（仅统计）
    OTHER_DATA = "other_data"  # 其他数据（仅统计）


class ScanStatus(str, Enum):
    """扫描状态枚举"""

    IDLE = "idle"  # 空闲
    SCANNING = "scanning"  # 扫描中
    COMPLETED = "completed"  # 完成
    FAILED = "failed"  # 失败


class CleanupStatus(str, Enum):
    """清理状态枚举"""

    PENDING = "pending"  # 等待中
    RUNNING = "running"  # 执行中
    COMPLETED = "completed"  # 完成
    FAILED = "failed"  # 失败


class FileInfo(BaseModel):
    """文件信息"""

    relative_path: str = Field(..., description="相对路径")
    name: str = Field(..., description="文件名")
    size: int = Field(..., description="文件大小（字节）")
    created_time: float = Field(..., description="创建时间（时间戳）")
    modified_time: float = Field(..., description="修改时间（时间戳）")
    chat_key: Optional[str] = Field(None, description="所属聊天key")
    plugin_key: Optional[str] = Field(None, description="所属插件key")


class ChatResourceInfo(BaseModel):
    """聊天资源信息"""

    chat_key: str = Field(..., description="聊天key")
    chat_name: Optional[str] = Field(None, description="聊天名称")
    total_size: int = Field(0, description="总大小（字节）")
    file_count: int = Field(0, description="文件数量")
    files: List[FileInfo] = Field(default_factory=list, description="文件列表")


class PluginResourceInfo(BaseModel):
    """插件资源信息"""

    plugin_key: str = Field(..., description="插件key")
    plugin_name: Optional[str] = Field(None, description="插件名称")
    total_size: int = Field(0, description="总大小（字节）")
    file_count: int = Field(0, description="文件数量")


class ResourceCategory(BaseModel):
    """资源分类"""

    resource_type: ResourceType = Field(..., description="资源类型")
    display_name: str = Field(..., description="显示名称（向后兼容）")
    description: str = Field(..., description="描述（向后兼容）")
    total_size: int = Field(0, description="总大小（字节）")
    file_count: int = Field(0, description="文件数量")
    can_cleanup: bool = Field(True, description="是否可清理")
    risk_level: str = Field("safe", description="风险等级: safe/warning/danger")
    risk_message: Optional[str] = Field(None, description="风险提示（向后兼容）")
    supports_time_filter: bool = Field(True, description="是否支持按时间过滤（False表示只能完全清理）")
    chat_resources: List[ChatResourceInfo] = Field(default_factory=list, description="按聊天分组的资源")
    plugin_resources: List[PluginResourceInfo] = Field(default_factory=list, description="按插件分组的资源")

    # i18n 扩展字段（可选，向后兼容，以 i18n_ 前缀便于字母排序聚合）
    i18n_description: Optional[I18nDict] = Field(None, description="描述国际化")
    i18n_display_name: Optional[I18nDict] = Field(None, description="显示名称国际化")
    i18n_risk_message: Optional[I18nDict] = Field(None, description="风险提示国际化")


class DiskInfo(BaseModel):
    """磁盘信息"""

    total_space: int = Field(..., description="系统总空间（字节）")
    used_space: int = Field(..., description="已使用空间（字节）")
    free_space: int = Field(..., description="可用空间（字节）")
    data_dir_size: int = Field(..., description="DATA_DIR占用空间（字节）")
    data_dir_path: str = Field(..., description="DATA_DIR路径")


class ScanSummary(BaseModel):
    """扫描摘要"""

    total_size: int = Field(0, description="总大小（字节）")
    total_files: int = Field(0, description="总文件数")
    start_time: Optional[datetime] = Field(None, description="开始时间")
    end_time: Optional[datetime] = Field(None, description="结束时间")
    duration_seconds: Optional[float] = Field(None, description="持续时长（秒）")
    scanned_categories: int = Field(0, description="扫描的分类数")


class ScanResult(BaseModel):
    """扫描结果"""

    scan_id: str = Field(..., description="扫描ID")
    status: ScanStatus = Field(..., description="扫描状态")
    disk_info: Optional[DiskInfo] = Field(None, description="磁盘信息")
    categories: List[ResourceCategory] = Field(default_factory=list, description="资源分类列表")
    summary: ScanSummary = Field(
        default_factory=lambda: ScanSummary(
            total_size=0,
            total_files=0,
            start_time=None,
            end_time=None,
            duration_seconds=None,
            scanned_categories=0,
        ),
        description="扫描摘要",
    )
    error_message: Optional[str] = Field(None, description="错误信息")


class ScanStatusResponse(BaseModel):
    """扫描状态响应"""

    status: ScanStatus = Field(..., description="扫描状态")
    progress: float = Field(0, description="进度（0-100）")
    current_category: Optional[str] = Field(None, description="当前扫描的分类")
    message: Optional[str] = Field(None, description="状态消息")


class CleanupRequest(BaseModel):
    """清理请求"""

    resource_types: List[ResourceType] = Field(..., description="要清理的资源类型")
    chat_keys: Optional[List[str]] = Field(None, description="要清理的聊天key列表")
    before_date: Optional[datetime] = Field(None, description="清理此日期之前的文件")
    dry_run: bool = Field(False, description="是否为模拟运行（不实际删除）")


class CleanupProgress(BaseModel):
    """清理进度"""

    task_id: str = Field(..., description="任务ID")
    status: CleanupStatus = Field(..., description="清理状态")
    progress: float = Field(0, description="进度（0-100）")
    processed_files: int = Field(0, description="已处理文件数")
    total_files: int = Field(0, description="总文件数")
    freed_space: int = Field(0, description="已释放空间（字节）")
    current_file: Optional[str] = Field(None, description="当前处理的文件")
    message: Optional[str] = Field(None, description="状态消息")


class CleanupResult(BaseModel):
    """清理结果"""

    task_id: str = Field(..., description="任务ID")
    status: CleanupStatus = Field(..., description="清理状态")
    total_files: int = Field(0, description="总文件数")
    deleted_files: int = Field(0, description="已删除文件数")
    failed_files: int = Field(0, description="删除失败文件数")
    freed_space: int = Field(0, description="释放的空间（字节）")
    start_time: Optional[datetime] = Field(None, description="开始时间")
    end_time: Optional[datetime] = Field(None, description="结束时间")
    duration_seconds: Optional[float] = Field(None, description="持续时长（秒）")
    error_message: Optional[str] = Field(None, description="错误信息")
    failed_file_list: List[str] = Field(default_factory=list, description="删除失败的文件列表")


# ============================================================
# 简单响应模型（避免返回无约束字典）
# ============================================================


class ScanStartResponse(BaseModel):
    """扫描启动响应"""

    scan_id: str = Field(..., description="扫描任务ID")


class CleanupStartResponse(BaseModel):
    """清理启动响应"""

    task_id: str = Field(..., description="清理任务ID")


class ScanProgressResponse(BaseModel):
    """扫描进度响应"""

    status: str = Field(..., description="扫描状态")
    progress: float = Field(0, description="进度（0-100）")
    message: Optional[str] = Field(None, description="状态消息")
